ccl/sqlproxyccl/tenant: TestResume failed

ccl/sqlproxyccl/tenant.TestResume [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9499596?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9499596?buildTab=artifacts#/) on release-23.1 @ [de239a7438f44d382c9aefceb65d9c39911dabd2](https://github.com/cockroachdb/cockroach/commits/de239a7438f44d382c9aefceb65d9c39911dabd2):


```
=== RUN   TestResume
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume311932878
    directory_cache_test.go:511: starting tenant 40
    directory_cache_test.go:267: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:267
        	Error:      	Condition never satisfied
        	Test:       	TestResume
    directory_cache_test.go:259: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:259
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
        	Test:       	TestResume
    directory_cache_test.go:259: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:259
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:259: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:259
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:259: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:259
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:259: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:259
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume311932878
--- FAIL: TestResume (19.91s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100682 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-master]
</p>
</details>
/cc @cockroachdb/sql-sessions @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResume.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-26723

ccl/sqlproxyccl/tenant.TestResume [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9600396?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9600396?buildTab=artifacts#/) on release-23.1 @ [ad16885ca3b4567ed5eb34646fe8281fd2d740e3](https://github.com/cockroachdb/cockroach/commits/ad16885ca3b4567ed5eb34646fe8281fd2d740e3):


```
=== RUN   TestResume
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume1505824533
    directory_cache_test.go:507: starting tenant 40
    directory_cache_test.go:263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:263
        	Error:      	Condition never satisfied
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume1505824533
--- FAIL: TestResume (19.14s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101429 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-release-23.1.0]
- #100682 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResume.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestResume [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9739345?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9739345?buildTab=artifacts#/) on release-23.1 @ [01f129a59ab66166d6ef7189a8930f2b37cced74](https://github.com/cockroachdb/cockroach/commits/01f129a59ab66166d6ef7189a8930f2b37cced74):


```
=== RUN   TestResume
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume3649493526
    directory_cache_test.go:507: starting tenant 40
    directory_cache_test.go:263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:263
        	Error:      	Condition never satisfied
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume3649493526
--- FAIL: TestResume (20.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101429 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-release-23.1.0]
- #100682 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResume.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestResume [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9892556?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9892556?buildTab=artifacts#/) on release-23.1 @ [fa8d237846052ae72cc44b3eaebe1f38c89834f8](https://github.com/cockroachdb/cockroach/commits/fa8d237846052ae72cc44b3eaebe1f38c89834f8):


```
=== RUN   TestResume
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume4039505146
    directory_cache_test.go:507: starting tenant 40
    directory_cache_test.go:263: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:263
        	Error:      	Condition never satisfied
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:255: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:255
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume4039505146
--- FAIL: TestResume (20.71s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101429 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-release-23.1.0]
- #100682 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResume.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestResume [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10523624?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10523624?buildTab=artifacts#/) on release-23.1 @ [407ad56a797df975f5cf5b0784dfe4a061cbf1c3](https://github.com/cockroachdb/cockroach/commits/407ad56a797df975f5cf5b0784dfe4a061cbf1c3):


```
=== RUN   TestResume
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume3991971400
    directory_cache_test.go:661: starting tenant 40
    directory_cache_test.go:442: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:442
        	Error:      	Condition never satisfied
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume3991971400
--- FAIL: TestResume (18.94s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #101429 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot todo-deprecate.branch-release-23.1.0]
- #100682 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResume.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/sqlproxyccl/tenant.TestResume [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10540886?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10540886?buildTab=artifacts#/) on release-23.1 @ [dec9fd64885d8df6dd23a7a3441de9e509637148](https://github.com/cockroachdb/cockroach/commits/dec9fd64885d8df6dd23a7a3441de9e509637148):


```
=== RUN   TestResume
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume2976050706
    directory_cache_test.go:661: starting tenant 40
    directory_cache_test.go:442: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:442
        	Error:      	Condition never satisfied
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	rpc error: code = Canceled desc = context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    directory_cache_test.go:434: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant_test/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:434
        	            				src/runtime/asm_amd64.s:1594
        	Error:      	Received unexpected error:
        	            	context canceled
        	Test:       	TestResume
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f215c8f665df7837d22d3ffe15cbecee/logTestResume2976050706
--- FAIL: TestResume (19.02s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #100682 ccl/sqlproxyccl/tenant: TestResume failed [C-test-failure O-robot T-serverless branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestResume.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/cliccl/mt_test_directory.go
+++ b/pkg/ccl/cliccl/mt_test_directory.go
@@ -9,14 +9,23 @@
 package cliccl
 
 import (
+	"bufio"
 	"context"
+	"crypto/tls"
 	"fmt"
 	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenantdirsvr"
 	"github.com/cockroachdb/cockroach/pkg/cli"
 	"github.com/cockroachdb/cockroach/pkg/cli/clierrorplus"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/spf13/cobra"
 )
 
@@ -53,7 +62,9 @@ func runDirectorySvr(cmd *cobra.Command, args []string) (returnErr error) {
 	}
 	defer stopper.Stop(ctx)
 
-	tds, err := tenantdirsvr.New(stopper, args...)
+	tds, err := tenantdirsvr.New(stopper, func(ctx context.Context, stopper *stop.Stopper, tenantID uint64) (string, error) {
+		return startTenant(ctx, stopper, tenantID, args)
+	})
 	if err != nil {
 		return err
 	}
@@ -67,3 +78,95 @@ func runDirectorySvr(cmd *cobra.Command, args []string) (returnErr error) {
 	stopper.AddCloser(stop.CloserFn(func() { _ = listenPort.Close() }))
 	return tds.Serve(listenPort)
 }
+
+func startTenant(
+	ctx context.Context, stopper *stop.Stopper, tenantID uint64, args []string,
+) (sqlAddress string, err error) {
+	// A hackish way to have the sql tenant process listen on known ports.
+	sqlListener, err := net.Listen("tcp", "")
+	if err != nil {
+		return "", err
+	}
+	httpListener, err := net.Listen("tcp", "")
+	if err != nil {
+		return "", err
+	}
+
+	if len(args) == 0 {
+		cockroachExecutable, err := os.Executable()
+		if err != nil {
+			return "", err
+		}
+		args = []string{
+			cockroachExecutable, "mt", "start-sql", "--kv-addrs=:26257", "--insecure",
+		}
+	}
+	args = append(args,
+		fmt.Sprintf("--sql-addr=%s", sqlListener.Addr().String()),
+		fmt.Sprintf("--http-addr=%s", httpListener.Addr().String()),
+		fmt.Sprintf("--tenant-id=%d", tenantID),
+	)
+	if err = sqlListener.Close(); err != nil {
+		return "", err
+	}
+	if err = httpListener.Close(); err != nil {
+		return "", err
+	}
+
+	c := exec.Command(args[0], args[1:]...)
+	c.Env = append(os.Environ(), "COCKROACH_TRUST_CLIENT_PROVIDED_SQL_REMOTE_ADDR=true")
+	var f writerFunc = func(p []byte) (int, error) {
+		sc := bufio.NewScanner(strings.NewReader(string(p)))
+		for sc.Scan() {
+			log.Infof(ctx, "%s", sc.Text())
+		}
+		return len(p), nil
+	}
+	c.Stdout = f
+	c.Stderr = f
+	err = c.Start()
+	if err != nil {
+		return "", err
+	}
+	stopper.AddCloser(stop.CloserFn(func() {
+		_ = c.Process.Kill()
+	}))
+	err = stopper.RunAsyncTask(ctx, "cmd-wait", func(ctx context.Context) {
+		if err := c.Wait(); err != nil {
+			log.Infof(ctx, "finished %s with err %s", c.Args, err)
+			return
+		}
+		log.Infof(ctx, "finished %s with success", c.Args)
+		stopper.Stop(ctx)
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// Wait for the tenant to show healthy
+	start := timeutil.Now()
+	transport := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	client := &http.Client{Transport: transport}
+	for {
+		time.Sleep(300 * time.Millisecond)
+		resp, err := client.Get(fmt.Sprintf("https://%s/health", httpListener.Addr().String()))
+		waitTime := timeutil.Since(start)
+		if err == nil {
+			resp.Body.Close()
+			log.Infof(ctx, "tenant is healthy")
+			break
+		}
+		if waitTime > 5*time.Second {
+			log.Infof(ctx, "waited more than 5 sec for the tenant to get healthy and it still isn't")
+			break
+		}
+		log.Infof(ctx, "waiting %s for healthy tenant: %s", waitTime, err)
+	}
+	return sqlAddress, nil
+}
+
+type writerFunc func(p []byte) (int, error)
+
+func (wf writerFunc) Write(p []byte) (int, error) { return wf(p) }

--- a/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go
+++ b/pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go
@@ -437,22 +437,21 @@ func TestResume(t *testing.T) {
 	}
 
 	// Eventually the tenant process will be resumed.
-	var processes map[net.Addr]*tenantdirsvr.Process
+	var sqlAddr string
 	testutils.SucceedsSoon(t, func() error {
-		processes = tds.Get(tenantID)
+		processes := tds.Get(tenantID)
 		if len(processes) != 1 {
 			return errors.Newf("expected 1 processes found %d", len(processes))
 		}
+		sqlAddr = processes[0].SQLAddr
 		return nil
 	})
 
 	// Wait until background goroutines complete.
 	wait.Wait()
 
-	for addr := range processes {
-		for i := 0; i < lookupCount; i++ {
-			require.Equal(t, addr.String(), addrs[i])
-		}
+	for i := 0; i < lookupCount; i++ {
+		require.Equal(t, sqlAddr, addrs[i])
 	}
 }
 
@@ -609,9 +608,8 @@ func destroyTenant(tc serverutils.TestClusterInterface, id roachpb.TenantID) err
 }
 
 func startTenant(
-	ctx context.Context, srv serverutils.TestServerInterface, id uint64,
-) (*tenantdirsvr.Process, error) {
-	tenantStopper := tenantdirsvr.NewSubStopper(srv.Stopper())
+	ctx context.Context, tenantStopper *stop.Stopper, srv serverutils.TestServerInterface, id uint64,
+) (string, error) {
 	t, err := srv.StartTenant(
 		ctx,
 		base.TestTenantArgs{
@@ -625,15 +623,11 @@ func startTenant(
 	if err != nil {
 		// Remap tenant "not found" error to GRPC NotFound error.
 		if testutils.IsError(err, "not found|no tenant found") {
-			return nil, status.Errorf(codes.NotFound, "tenant %d not found", id)
+			return "", status.Errorf(codes.NotFound, "tenant %d not found", id)
 		}
-		return nil, err
+		return "", err
 	}
-	sqlAddr, err := net.ResolveTCPAddr("tcp", t.SQLAddr())
-	if err != nil {
-		return nil, err
-	}
-	return &tenantdirsvr.Process{SQL: sqlAddr, Stopper: tenantStopper}, nil
+	return t.SQLAddr(), nil
 }
 
 // Setup directory cache that uses a client connected to a test directory server
@@ -657,17 +651,16 @@ func newTestDirectoryCache(
 	})
 	clusterStopper := tc.Stopper()
 	var err error
-	tds, err = tenantdirsvr.New(clusterStopper)
-	require.NoError(t, err)
-	tds.TenantStarterFunc = func(ctx context.Context, tenantID uint64) (*tenantdirsvr.Process, error) {
+	tds, err = tenantdirsvr.New(clusterStopper, func(ctx context.Context, stopper *stop.Stopper, tenantID uint64) (string, error) {
 		t.Logf("starting tenant %d", tenantID)
-		process, err := startTenant(ctx, tc.Server(0), tenantID)
+		sqlAddr, err := startTenant(ctx, stopper, tc.Server(0), tenantID)
 		if err != nil {
-			return nil, err
+			return "", err
 		}
 		t.Logf("tenant %d started", tenantID)
-		return process, nil
-	}
+		return sqlAddr, nil
+	})
+	require.NoError(t, err)
 
 	listenPort, err := net.Listen("tcp", "127.0.0.1:0")
 	require.NoError(t, err)

--- a/pkg/ccl/sqlproxyccl/tenantdirsvr/test_directory_svr.go
+++ b/pkg/ccl/sqlproxyccl/tenantdirsvr/test_directory_svr.go
@@ -9,21 +9,13 @@
 package tenantdirsvr
 
 import (
-	"bufio"
 	"container/list"
 	"context"
-	"crypto/tls"
-	"fmt"
 	"net"
-	"net/http"
 	"os"
-	"os/exec"
-	"strings"
-	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/tenant"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
-	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
@@ -35,13 +27,6 @@ import (
 
 // Make sure that TestDirectoryServer implements the DirectoryServer interface.
 var _ tenant.DirectoryServer = (*TestDirectoryServer)(nil)
-
-// Process stores information about a running tenant process.
-type Process struct {
-	Stopper *stop.Stopper
-	Cmd     *exec.Cmd
-	SQL     net.Addr
-}
 
 // NewSubStopper creates a new stopper that will be stopped when either the
 // parent is stopped or its own Stop is called. The code is slightly more
@@ -67,44 +52,46 @@ func NewSubStopper(parentStopper *stop.Stopper) *stop.Stopper {
 	return subStopper
 }
 
+// Process stores information about a running tenant process.
+type Process struct {
+	Stopper *stop.Stopper
+	SQLAddr string
+}
+
 // TestDirectoryServer is a directory server implementation that is used for
 // testing.
 type TestDirectoryServer struct {
-	args                []string
 	stopper             *stop.Stopper
 	grpcServer          *grpc.Server
 	cockroachExecutable string
-	// TenantStarterFunc will be used to launch a new tenant process.
-	TenantStarterFunc func(ctx context.Context, tenantID uint64) (*Process, error)
+	// tenantStarterFunc will be used to launch a new tenant process.
+	tenantStarterFunc TenantStarterFunc
 
-	// When both mutexes need to be held, the locking should always be proc
-	// first and listen second.
-	proc struct {
-		syncutil.RWMutex
-		processByAddrByTenantID map[uint64]map[net.Addr]*Process
-	}
-	listen struct {
-		syncutil.RWMutex
-		eventListeners *list.List
+	mu struct {
+		syncutil.Mutex
+		// A map of tenantIDs to sql server addresses
+		processByAddrByTenantID map[uint64][]Process
+		eventListeners          *list.List
 	}
 }
 
+type TenantStarterFunc func(ctx context.Context, stopper *stop.Stopper, tenantID uint64) (string, error)
+
 // New will create a new server.
-func New(stopper *stop.Stopper, args ...string) (*TestDirectoryServer, error) {
+func New(stopper *stop.Stopper, createTenant TenantStarterFunc) (*TestDirectoryServer, error) {
 	// Determine the path to cockroach executable.
 	cockroachExecutable, err := os.Executable()
 	if err != nil {
 		return nil, err
 	}
 	dir := &TestDirectoryServer{
-		args:                args,
 		grpcServer:          grpc.NewServer(),
 		stopper:             stopper,
 		cockroachExecutable: cockroachExecutable,
 	}
-	dir.TenantStarterFunc = dir.startTenantLocked
-	dir.proc.processByAddrByTenantID = map[uint64]map[net.Addr]*Process{}
-	dir.listen.eventListeners = list.New()
+	dir.tenantStarterFunc = createTenant
+	dir.mu.processByAddrByTenantID = map[uint64][]Process{}
+	dir.mu.eventListeners = list.New()
 	stopper.AddCloser(stop.CloserFn(dir.grpcServer.GracefulStop))
 	tenant.RegisterDirectoryServer(dir.grpcServer, dir)
 	return dir, nil
@@ -117,16 +104,10 @@ func (s *TestDirectoryServer) Stopper() *stop.Stopper {
 }
 
 // Get a tenant's list of pods and the process information for each pod.
-func (s *TestDirectoryServer) Get(id roachpb.TenantID) (result map[net.Addr]*Process) {
-	result = make(map[net.Addr]*Process)
-	s.proc.RLock()
-	defer s.proc.RUnlock()
-	processes, ok := s.proc.processByAddrByTenantID[id.ToUint64()]
-	if ok {
-		for k, v := range processes {
-			result[k] = v
-		}
-	}
+func (s *TestDirectoryServer) Get(id roachpb.TenantID) (result []Process) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	result = append(result, s.mu.processByAddrByTenantID[id.ToUint64()]...)
 	return
 }
 
@@ -146,17 +127,24 @@ func (s *TestDirectoryServer) StartTenant(ctx context.Context, id roachpb.Tenant
 
 	ctx = logtags.AddTag(ctx, "tenant", id)
 
-	s.proc.Lock()
-	defer s.proc.Unlock()
+	tenantStopper := NewSubStopper(s.stopper)
 
-	process, err := s.TenantStarterFunc(ctx, id.ToUint64())
+	sqlAddr, err := s.tenantStarterFunc(ctx, tenantStopper, id.ToUint64())
 	if err != nil {
 		return err
 	}
 
-	s.registerInstanceLocked(id.ToUint64(), process)
-	process.Stopper.AddCloser(stop.CloserFn(func() {
-		s.deregisterInstance(id.ToUint64(), process.SQL)
+	func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.registerInstanceLocked(id.ToUint64(), Process{
+			Stopper: tenantStopper,
+			SQLAddr: sqlAddr,
+		})
+	}()
+
+	tenantStopper.AddCloser(stop.CloserFn(func() {
+		s.deregisterInstance(id.ToUint64(), sqlAddr)
 	}))
 
 	return nil
@@ -182,8 +170,8 @@ func (s *TestDirectoryServer) ListPods(
 	ctx context.Context, req *tenant.ListPodsRequest,
 ) (*tenant.ListPodsResponse, error) {
 	ctx = logtags.AddTag(ctx, "tenant", req.TenantID)
-	s.proc.RLock()
-	defer s.proc.RUnlock()
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	return s.listLocked(ctx, req)
 }
 
@@ -200,9 +188,9 @@ func (s *TestDirectoryServer) WatchPods(
 	// Make the channel with a small buffer to allow for a burst of notifications
 	// and a slow receiver.
 	c := make(chan *tenant.WatchPodsResponse, 10)
-	s.listen.Lock()
-	elem := s.listen.eventListeners.PushBack(c)
-	s.listen.Unlock()
+	s.mu.Lock()
+	elem := s.mu.eventListeners.PushBack(c)
+	s.mu.Unlock()
 	err := s.stopper.RunTask(context.Background(), "watch-pods-server",
 		func(ctx context.Context) {
 		out:
@@ -213,17 +201,17 @@ func (s *TestDirectoryServer) WatchPods(
 						break out
 					}
 					if err := server.Send(e); err != nil {
-						s.listen.Lock()
-						s.listen.eventListeners.Remove(elem)
+						s.mu.Lock()
+						s.mu.eventListeners.Remove(elem)
 						close(c)
-						s.listen.Unlock()
+						s.mu.Unlock()
 						break out
 					}
 				case <-s.stopper.ShouldQuiesce():
-					s.listen.Lock()
-					s.listen.eventListeners.Remove(elem)
+					s.mu.Lock()
+					s.mu.eventListeners.Remove(elem)
 					close(c)
-					s.listen.Unlock()
+					s.mu.Unlock()
 					break out
 				}
 			}
@@ -248,17 +236,15 @@ func (s *TestDirectoryServer) WatchTenants(
 // test directory. This causes the proxy to start enforcing short idle
 // connection timeouts in order to drain the connections to the pod.
 func (s *TestDirectoryServer) Drain() {
-	s.proc.RLock()
-	defer s.proc.RUnlock()
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-	for tenantID, processByAddr := range s.proc.processByAddrByTenantID {
-		for addr := range processByAddr {
-			s.listen.RLock()
-			defer s.listen.RUnlock()
+	for tenantID, processByAddr := range s.mu.processByAddrByTenantID {
+		for _, process := range processByAddr {
 			s.notifyEventListenersLocked(&tenant.WatchPodsResponse{
 				Pod: &tenant.Pod{
 					TenantID:       tenantID,
-					Addr:           addr.String(),
+					Addr:           process.SQLAddr,
 					State:          tenant.DRAINING,
 					StateTimestamp: timeutil.Now(),
 				},
@@ -268,7 +254,7 @@ func (s *TestDirectoryServer) Drain() {
 }
 
 func (s *TestDirectoryServer) notifyEventListenersLocked(req *tenant.WatchPodsResponse) {
-	for e := s.listen.eventListeners.Front(); e != nil; {
+	for e := s.mu.eventListeners.Front(); e != nil; {
 		select {
 		case e.Value.(chan *tenant.WatchPodsResponse) <- req:
 			e = e.Next()
@@ -278,7 +264,7 @@ func (s *TestDirectoryServer) notifyEventListenersLocked(req *tenant.WatchPodsRe
 			eToClose := e
 			e = e.Next()
 			close(eToClose.Value.(chan *tenant.WatchPodsResponse))
-			s.listen.eventListeners.Remove(eToClose)
+			s.mu.eventListeners.Remove(eToClose)
 		}
 	}
 }
@@ -297,21 +283,36 @@ func (s *TestDirectoryServer) EnsurePod(
 
 	ctx = logtags.AddTag(ctx, "tenant", req.TenantID)
 
-	s.proc.Lock()
-	defer s.proc.Unlock()
-
-	lst, err := s.listLocked(ctx, &tenant.ListPodsRequest{TenantID: req.TenantID})
+	tenantStopper, sqlAddr, err := func() (*stop.Stopper, string, error) {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		lst, err := s.listLocked(ctx, &tenant.ListPodsRequest{TenantID: req.TenantID})
+		if err != nil {
+			return nil, "", err
+		}
+		if len(lst.Pods) != 0 {
+			return nil, "", nil
+		}
+		tenantStopper := NewSubStopper(s.stopper)
+		sqlAddr, err := s.tenantStarterFunc(ctx, tenantStopper, req.TenantID)
+		if err != nil {
+			return nil, "", err
+		}
+		s.registerInstanceLocked(req.TenantID, Process{
+			Stopper: tenantStopper,
+			SQLAddr: sqlAddr,
+		})
+		return tenantStopper, sqlAddr, nil
+	}()
 	if err != nil {
 		return nil, err
 	}
-	if len(lst.Pods) == 0 {
-		process, err := s.TenantStarterFunc(ctx, req.TenantID)
-		if err != nil {
-			return nil, err
-		}
-		s.registerInstanceLocked(req.TenantID, process)
-		process.Stopper.AddCloser(stop.CloserFn(func() {
-			s.deregisterInstance(req.TenantID, process.SQL)
+
+	if tenantStopper != nil {
+		// Registering a stopper closer that removes the instance must occur
+		// outside of the mutex.
+		tenantStopper.AddCloser(stop.CloserFn(func() {
+			s.deregisterInstance(req.TenantID, sqlAddr)
 		}))
 	}
 
@@ -326,15 +327,15 @@ func (s *TestDirectoryServer) Serve(listener net.Listener) error {
 func (s *TestDirectoryServer) listLocked(
 	_ context.Context, req *tenant.ListPodsRequest,
 ) (*tenant.ListPodsResponse, error) {
-	processByAddr, ok := s.proc.processByAddrByTenantID[req.TenantID]
+	processByAddr, ok := s.mu.processByAddrByTenantID[req.TenantID]
 	if !ok {
 		return &tenant.ListPodsResponse{}, nil
 	}
 	resp := tenant.ListPodsResponse{}
-	for addr := range processByAddr {
+	for _, process := range processByAddr {
 		resp.Pods = append(resp.Pods, &tenant.Pod{
 			TenantID:       req.TenantID,
-			Addr:           addr.String(),
+			Addr:           process.SQLAddr,
 			State:          tenant.RUNNING,
 			StateTimestamp: timeutil.Now(),
 		})
@@ -342,144 +343,42 @@ func (s *TestDirectoryServer) listLocked(
 	return &resp, nil
 }
 
-func (s *TestDirectoryServer) registerInstanceLocked(tenantID uint64, process *Process) {
-	processByAddr, ok := s.proc.processByAddrByTenantID[tenantID]
-	if !ok {
-		processByAddr = map[net.Addr]*Process{}
-		s.proc.processByAddrByTenantID[tenantID] = processByAddr
-	}
-	processByAddr[process.SQL] = process
-
-	s.listen.RLock()
-	defer s.listen.RUnlock()
+func (s *TestDirectoryServer) registerInstanceLocked(tenantID uint64, process Process) {
+	s.mu.processByAddrByTenantID[tenantID] = append(s.mu.processByAddrByTenantID[tenantID], process)
 	s.notifyEventListenersLocked(&tenant.WatchPodsResponse{
 		Pod: &tenant.Pod{
 			TenantID:       tenantID,
-			Addr:           process.SQL.String(),
+			Addr:           process.SQLAddr,
 			State:          tenant.RUNNING,
 			StateTimestamp: timeutil.Now(),
 		},
 	})
 }
 
-func (s *TestDirectoryServer) deregisterInstance(tenantID uint64, sql net.Addr) {
-	s.proc.Lock()
-	defer s.proc.Unlock()
-	processByAddr, ok := s.proc.processByAddrByTenantID[tenantID]
-	if !ok {
-		return
-	}
+func (s *TestDirectoryServer) deregisterInstance(tenantID uint64, sqlAddr string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	processByAddr := s.mu.processByAddrByTenantID[tenantID]
 
-	if _, ok = processByAddr[sql]; ok {
-		delete(processByAddr, sql)
-
-		s.listen.RLock()
-		defer s.listen.RUnlock()
-		s.notifyEventListenersLocked(&tenant.WatchPodsResponse{
-			Pod: &tenant.Pod{
-				TenantID:       tenantID,
-				Addr:           sql.String(),
-				State:          tenant.DELETING,
-				StateTimestamp: timeutil.Now(),
-			},
-		})
-	}
-}
-
-type writerFunc func(p []byte) (int, error)
-
-func (wf writerFunc) Write(p []byte) (int, error) { return wf(p) }
-
-// startTenantLocked is the default tenant process startup logic that runs the
-// cockroach db executable out of process.
-func (s *TestDirectoryServer) startTenantLocked(
-	ctx context.Context, tenantID uint64,
-) (*Process, error) {
-	// A hackish way to have the sql tenant process listen on known ports.
-	sqlListener, err := net.Listen("tcp", "")
-	if err != nil {
-		return nil, err
-	}
-	httpListener, err := net.Listen("tcp", "")
-	if err != nil {
-		return nil, err
-	}
-	process := &Process{SQL: sqlListener.Addr()}
-	args := s.args
-	if len(args) == 0 {
-		args = append(args,
-			s.cockroachExecutable, "mt", "start-sql", "--kv-addrs=:26257", "--insecure",
-		)
-	}
-	args = append(args,
-		fmt.Sprintf("--sql-addr=%s", sqlListener.Addr().String()),
-		fmt.Sprintf("--http-addr=%s", httpListener.Addr().String()),
-		fmt.Sprintf("--tenant-id=%d", tenantID),
-	)
-	if err = sqlListener.Close(); err != nil {
-		return nil, err
-	}
-	if err = httpListener.Close(); err != nil {
-		return nil, err
-	}
-
-	c := exec.Command(args[0], args[1:]...)
-	process.Cmd = c
-	c.Env = append(os.Environ(), "COCKROACH_TRUST_CLIENT_PROVIDED_SQL_REMOTE_ADDR=true")
-	var f writerFunc = func(p []byte) (int, error) {
-		sc := bufio.NewScanner(strings.NewReader(string(p)))
-		for sc.Scan() {
-			log.Infof(ctx, "%s", sc.Text())
+	var remaining []Process
+	for _, process := range processByAddr {
+		if sqlAddr == process.SQLAddr {
+			s.notifyEventListenersLocked(&tenant.WatchPodsResponse{
+				Pod: &tenant.Pod{
+					TenantID:       tenantID,
+					Addr:           sqlAddr,
+					State:          tenant.DELETING,
+					StateTimestamp: timeutil.Now(),
+				},
+			})
+		} else {
+			remaining = append(remaining, process)
 		}
-		return len(p), nil
-	}
-	c.Stdout = f
-	c.Stderr = f
-	err = c.Start()
-	if err != nil {
-		return nil, err
-	}
-	process.Stopper = NewSubStopper(s.stopper)
-	process.Stopper.AddCloser(stop.CloserFn(func() {
-		_ = c.Process.Kill()
-		s.deregisterInstance(tenantID, process.SQL)
-	}))
-	err = s.stopper.RunAsyncTask(ctx, "cmd-wait", func(ctx context.Context) {
-		if err := c.Wait(); err != nil {
-			log.Infof(ctx, "finished %s with err %s", process.Cmd.Args, err)
-			return
-		}
-		log.Infof(ctx, "finished %s with success", process.Cmd.Args)
-		process.Stopper.Stop(ctx)
-	})
-	if err != nil {
-		return nil, err
 	}
 
-	// Wait for the tenant to show healthy
-	start := timeutil.Now()
-	transport := &http.Transport{
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	if len(remaining) == 0 {
+		delete(s.mu.processByAddrByTenantID, tenantID)
+	} else {
+		s.mu.processByAddrByTenantID[tenantID] = remaining
 	}
-	client := &http.Client{Transport: transport}
-	for {
-		time.Sleep(300 * time.Millisecond)
-		resp, err := client.Get(fmt.Sprintf("https://%s/health", httpListener.Addr().String()))
-		waitTime := timeutil.Since(start)
-		if err == nil {
-			resp.Body.Close()
-			log.Infof(ctx, "tenant is healthy")
-			break
-		}
-		if waitTime > 5*time.Second {
-			log.Infof(ctx, "waited more than 5 sec for the tenant to get healthy and it still isn't")
-			break
-		}
-		log.Infof(ctx, "waiting %s for healthy tenant: %s", waitTime, err)
-	}
-
-	// We currently set the password when we spawn the first tenant so wait
-	// a second to ensure the password is set.
-	time.Sleep(time.Second)
-	return process, nil
 }

